File: qa/saleads_mi_negocio_full_test.py
#!/usr/bin/env python3
"""
End-to-end validation for the SaleADS.ai "Mi Negocio" workflow.

This script is environment-agnostic:
- It never hardcodes a SaleADS domain.
- It can run against any environment URL passed at runtime.
"""

from __future__ import annotations

import argparse
import json
import os
import re
from datetime import datetime, timezone
from pathlib import Path
from typing import Callable

from playwright.sync_api import BrowserContext, Page, TimeoutError as PlaywrightTimeoutError, sync_playwright


class SaleadsMiNegocioFullTest:
    ACCOUNT_EMAIL = "juanlucasbarbiergarzon@gmail.com"

    def __init__(self, page: Page, context: BrowserContext, output_dir: Path) -> None:
        self.page = page
        self.context = context
        self.output_dir = output_dir
        self.results: dict[str, dict[str, str]] = {}
        self.legal_urls: dict[str, str] = {}

    @staticmethod
    def _text_regex(text: str) -> re.Pattern[str]:
        return re.compile(rf"{re.escape(text)}", re.IGNORECASE)

    def _timestamped_name(self, label: str) -> str:
        safe_label = re.sub(r"[^a-zA-Z0-9_-]+", "_", label).strip("_").lower()
        stamp = datetime.now(timezone.utc).strftime("%Y%m%dT%H%M%SZ")
        return f"{stamp}_{safe_label}.png"

    def _wait_ui_ready(self, target_page: Page | None = None) -> None:
        page = target_page or self.page
        page.wait_for_load_state("domcontentloaded")
        try:
            page.wait_for_load_state("networkidle", timeout=7000)
        except PlaywrightTimeoutError:
            # Some environments keep long-lived connections; DOM readiness is enough.
            pass
        page.wait_for_timeout(750)

    def _try_click_by_text(self, texts: list[str]) -> bool:
        locators = []
        for text in texts:
            regex = self._text_regex(text)
            locators.extend(
                [
                    self.page.get_by_role("button", name=regex).first,
                    self.page.get_by_role("link", name=regex).first,
                    self.page.get_by_role("menuitem", name=regex).first,
                    self.page.get_by_role("tab", name=regex).first,
                    self.page.get_by_text(regex).first,
                ]
            )

        for locator in locators:
            try:
                locator.wait_for(state="visible", timeout=1200)
                locator.click()
                self._wait_ui_ready()
                return True
            except PlaywrightTimeoutError:
                continue
        return False

    def _assert_visible_text(self, text: str, timeout: int = 8000) -> None:
        self.page.get_by_text(self._text_regex(text)).first.wait_for(state="visible", timeout=timeout)

    def _capture(self, label: str, full_page: bool = False, target_page: Page | None = None) -> str:
        self.output_dir.mkdir(parents=True, exist_ok=True)
        screenshot_path = self.output_dir / self._timestamped_name(label)
        (target_page or self.page).screenshot(path=str(screenshot_path), full_page=full_page)
        return str(screenshot_path)

    def _record_step(self, key: str, name: str, passed: bool, details: str) -> None:
        self.results[key] = {
            "name": name,
            "status": "PASS" if passed else "FAIL",
            "details": details,
        }

    def _run_step(self, key: str, name: str, action: Callable[[], str]) -> None:
        try:
            details = action()
            self._record_step(key, name, True, details)
        except Exception as exc:  # noqa: BLE001 - we want full step-level reporting.
            self._record_step(key, name, False, str(exc))

    def step_1_login(self) -> str:
        self._wait_ui_ready()

        clicked = self._try_click_by_text(
            [
                "Sign in with Google",
                "Iniciar sesión con Google",
                "Continuar con Google",
                "Login with Google",
                "Google",
            ]
        )
        if not clicked:
            raise AssertionError("No se encontró un botón/enlace de login con Google.")

        # If Google account selector appears, choose the requested account.
        if self.page.get_by_text(self._text_regex(self.ACCOUNT_EMAIL)).first.is_visible(timeout=4000):
            self.page.get_by_text(self._text_regex(self.ACCOUNT_EMAIL)).first.click()
            self._wait_ui_ready()
        else:
            # Fallback: account chooser may appear in a popup.
            for popup in self.context.pages:
                if popup is self.page:
                    continue
                try:
                    popup.bring_to_front()
                    popup.get_by_text(self._text_regex(self.ACCOUNT_EMAIL)).first.click(timeout=4000)
                    self._wait_ui_ready(target_page=popup)
                    break
                except Exception:
                    continue
                finally:
                    self.page.bring_to_front()

        # Validate app interface and left sidebar.
        sidebar_found = False
        sidebar_locators = [
            self.page.get_by_role("navigation").first,
            self.page.locator("aside").first,
            self.page.get_by_text(self._text_regex("Mi Negocio")).first,
            self.page.get_by_text(self._text_regex("Negocio")).first,
        ]
        for locator in sidebar_locators:
            try:
                locator.wait_for(state="visible", timeout=7000)
                sidebar_found = True
                break
            except PlaywrightTimeoutError:
                continue

        if not sidebar_found:
            raise AssertionError("No se detectó interfaz principal con sidebar visible tras login.")

        screenshot = self._capture("step1_dashboard_loaded")
        return f"Dashboard y sidebar visibles. Screenshot: {screenshot}"

    def step_2_open_mi_negocio_menu(self) -> str:
        self._wait_ui_ready()

        # Try opening parent "Negocio" section first, then "Mi Negocio".
        self._try_click_by_text(["Negocio"])
        clicked_mi_negocio = self._try_click_by_text(["Mi Negocio"])
        if not clicked_mi_negocio:
            raise AssertionError("No se pudo hacer click en 'Mi Negocio'.")

        self._assert_visible_text("Agregar Negocio")
        self._assert_visible_text("Administrar Negocios")

        screenshot = self._capture("step2_menu_expanded")
        return f"Submenú de Mi Negocio expandido y validado. Screenshot: {screenshot}"

    def step_3_validate_agregar_negocio_modal(self) -> str:
        clicked = self._try_click_by_text(["Agregar Negocio"])
        if not clicked:
            raise AssertionError("No se pudo abrir 'Agregar Negocio'.")

        self._assert_visible_text("Crear Nuevo Negocio")
        self._assert_visible_text("Nombre del Negocio")
        self._assert_visible_text("Tienes 2 de 3 negocios")
        self._assert_visible_text("Cancelar")
        self._assert_visible_text("Crear Negocio")

        # Optional actions requested by the flow.
        name_field = self.page.get_by_label(self._text_regex("Nombre del Negocio")).first
        if not name_field.is_visible(timeout=2500):
            name_field = self.page.get_by_placeholder(self._text_regex("Nombre del Negocio")).first
        name_field.click()
        name_field.fill("Negocio Prueba Automatización")
        self._wait_ui_ready()

        screenshot = self._capture("step3_agregar_negocio_modal")
        self._try_click_by_text(["Cancelar"])
        return f"Modal validado y cerrado. Screenshot: {screenshot}"

    def step_4_open_administrar_negocios(self) -> str:
        # Re-open menu in case it collapsed after modal close.
        self._try_click_by_text(["Mi Negocio"])

        clicked = self._try_click_by_text(["Administrar Negocios"])
        if not clicked:
            raise AssertionError("No se pudo acceder a 'Administrar Negocios'.")

        self._assert_visible_text("Información General")
        self._assert_visible_text("Detalles de la Cuenta")
        self._assert_visible_text("Tus Negocios")
        self._assert_visible_text("Sección Legal")

        screenshot = self._capture("step4_administrar_negocios_view", full_page=True)
        return f"Vista de cuenta validada. Screenshot: {screenshot}"

    def step_5_validate_informacion_general(self) -> str:
        self._assert_visible_text("Información General")
        self._assert_visible_text("BUSINESS PLAN")
        self._assert_visible_text("Cambiar Plan")

        full_text = self.page.inner_text("body")
        email_match = re.search(r"[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}", full_text)
        if not email_match:
            raise AssertionError("No se encontró email visible en 'Información General'.")

        lines = [line.strip() for line in full_text.splitlines() if line.strip()]
        ignored = {"información general", "business plan", "cambiar plan"}
        likely_name = next(
            (
                line
                for line in lines
                if "@" not in line
                and line.lower() not in ignored
                and len(line.split()) >= 2
                and len(line) <= 80
            ),
            None,
        )
        if not likely_name:
            raise AssertionError("No se detectó un nombre de usuario visible.")

        return f"Nombre visible: '{likely_name}'. Email visible: '{email_match.group(0)}'."

    def step_6_validate_detalles_cuenta(self) -> str:
        self._assert_visible_text("Detalles de la Cuenta")
        self._assert_visible_text("Cuenta creada")
        self._assert_visible_text("Estado activo")
        self._assert_visible_text("Idioma seleccionado")
        return "Se validaron 'Cuenta creada', 'Estado activo' e 'Idioma seleccionado'."

    def step_7_validate_tus_negocios(self) -> str:
        self._assert_visible_text("Tus Negocios")
        self._assert_visible_text("Agregar Negocio")
        self._assert_visible_text("Tienes 2 de 3 negocios")
        return "Lista de negocios y contador de límite validados."

    def _validate_legal_page(self, link_text: str, expected_heading: str, screenshot_label: str) -> str:
        app_url_before = self.page.url
        opened_new_tab = False
        target_page = self.page

        try:
            with self.context.expect_page(timeout=5000) as new_page_event:
                clicked = self._try_click_by_text([link_text])
                if not clicked:
                    raise AssertionError(f"No se pudo hacer click en '{link_text}'.")
            target_page = new_page_event.value
            opened_new_tab = True
        except PlaywrightTimeoutError:
            # Link click was performed but no new tab appeared, so it likely opened in same tab.
            target_page = self.page

        target_page.bring_to_front()
        self._wait_ui_ready(target_page=target_page)

        target_page.get_by_text(self._text_regex(expected_heading)).first.wait_for(state="visible", timeout=12000)
        legal_text = target_page.locator("main, body").inner_text()
        if len(re.sub(r"\s+", "", legal_text)) < 120:
            raise AssertionError(f"No se detectó contenido legal suficiente para '{link_text}'.")

        screenshot = self._capture(screenshot_label, full_page=True, target_page=target_page)
        self.legal_urls[link_text] = target_page.url

        if opened_new_tab:
            target_page.close()
            self.page.bring_to_front()
            self._wait_ui_ready()
        else:
            # Cleanup for same-tab behavior.
            if self.page.url != app_url_before:
                self.page.go_back()
                self._wait_ui_ready()

        return f"{link_text} validado. URL final: {self.legal_urls[link_text]}. Screenshot: {screenshot}"

    def step_8_validate_terminos(self) -> str:
        return self._validate_legal_page(
            link_text="Términos y Condiciones",
            expected_heading="Términos y Condiciones",
            screenshot_label="step8_terminos_y_condiciones",
        )

    def step_9_validate_privacidad(self) -> str:
        return self._validate_legal_page(
            link_text="Política de Privacidad",
            expected_heading="Política de Privacidad",
            screenshot_label="step9_politica_de_privacidad",
        )

    def run(self) -> dict[str, object]:
        self._run_step("login", "Login", self.step_1_login)
        self._run_step("mi_negocio_menu", "Mi Negocio menu", self.step_2_open_mi_negocio_menu)
        self._run_step("agregar_modal", "Agregar Negocio modal", self.step_3_validate_agregar_negocio_modal)
        self._run_step("administrar_view", "Administrar Negocios view", self.step_4_open_administrar_negocios)
        self._run_step("informacion_general", "Información General", self.step_5_validate_informacion_general)
        self._run_step("detalles_cuenta", "Detalles de la Cuenta", self.step_6_validate_detalles_cuenta)
        self._run_step("tus_negocios", "Tus Negocios", self.step_7_validate_tus_negocios)
        self._run_step("terminos", "Términos y Condiciones", self.step_8_validate_terminos)
        self._run_step("privacidad", "Política de Privacidad", self.step_9_validate_privacidad)

        report = {
            "test_name": "saleads_mi_negocio_full_test",
            "generated_at_utc": datetime.now(timezone.utc).isoformat(),
            "results": self.results,
            "legal_urls": self.legal_urls,
        }
        return report


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="SaleADS Mi Negocio full workflow validation.")
    parser.add_argument(
        "--url",
        default=os.getenv("SALEADS_START_URL", "").strip(),
        help="SaleADS login URL for the target environment (dev/staging/prod).",
    )
    parser.add_argument(
        "--artifacts-dir",
        default="qa/artifacts/saleads_mi_negocio_full_test",
        help="Directory where screenshots and report JSON will be stored.",
    )
    parser.add_argument(
        "--headless",
        action="store_true",
        help="Run browser in headless mode (default: headed).",
    )
    return parser.parse_args()


def main() -> int:
    args = parse_args()
    if not args.url:
        raise SystemExit("Missing --url (or SALEADS_START_URL env var).")

    artifacts_dir = Path(args.artifacts_dir)
    artifacts_dir.mkdir(parents=True, exist_ok=True)

    with sync_playwright() as playwright:
        browser = playwright.chromium.launch(headless=args.headless)
        context = browser.new_context()
        page = context.new_page()
        page.goto(args.url, wait_until="domcontentloaded")

        runner = SaleadsMiNegocioFullTest(page=page, context=context, output_dir=artifacts_dir)
        report = runner.run()

        report_path = artifacts_dir / "report.json"
        report_path.write_text(json.dumps(report, indent=2, ensure_ascii=False), encoding="utf-8")
        print(json.dumps(report, indent=2, ensure_ascii=False))
        print(f"\nReporte guardado en: {report_path}")

        context.close()
        browser.close()
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: qa/saleads-mi-negocio-full-test.mjs
import fs from "node:fs/promises";
import path from "node:path";
import { chromium } from "playwright";

const REPORT_FIELDS = [
  "Login",
  "Mi Negocio menu",
  "Agregar Negocio modal",
  "Administrar Negocios view",
  "Información General",
  "Detalles de la Cuenta",
  "Tus Negocios",
  "Términos y Condiciones",
  "Política de Privacidad",
];

const GOOGLE_ACCOUNT_EMAIL =
  process.env.GOOGLE_ACCOUNT_EMAIL || "juanlucasbarbiergarzon@gmail.com";
const ARTIFACTS_DIR = path.resolve(
  process.env.SALEADS_ARTIFACTS_DIR || "artifacts/saleads-mi-negocio",
);
const HEADLESS = parseBoolean(process.env.HEADLESS, true);

function parseBoolean(value, defaultValue) {
  if (value === undefined) return defaultValue;
  return ["1", "true", "yes", "on"].includes(String(value).toLowerCase());
}

function createFieldResult() {
  return {
    status: "PASS",
    checks: [],
    errors: [],
    evidence: [],
  };
}

const report = Object.fromEntries(
  REPORT_FIELDS.map((field) => [field, createFieldResult()]),
);
const metadata = {
  startedAt: new Date().toISOString(),
  googleAccountEmail: GOOGLE_ACCOUNT_EMAIL,
  urls: {},
  artifactsDir: ARTIFACTS_DIR,
};

function recordCheck(field, description, passed, details = "") {
  const target = report[field];
  target.checks.push({ description, passed, details });
  if (!passed) {
    target.status = "FAIL";
    target.errors.push(details || description);
  }
}

function recordPresence(field, description, present, missingDetails) {
  recordCheck(
    field,
    description,
    Boolean(present),
    present ? "" : missingDetails,
  );
}

function finalizeFieldStatuses() {
  for (const field of REPORT_FIELDS) {
    const target = report[field];
    if (target.checks.length === 0) {
      target.status = "FAIL";
      if (!target.errors.includes("Step not executed")) {
        target.errors.push("Step not executed");
      }
      continue;
    }

    const hasFailures = target.checks.some((check) => !check.passed);
    target.status = hasFailures ? "FAIL" : "PASS";
  }
}

async function ensureArtifactsDir() {
  await fs.mkdir(ARTIFACTS_DIR, { recursive: true });
}

async function waitForUi(page, timeout = 15000) {
  await page.waitForLoadState("domcontentloaded", { timeout }).catch(() => {});
  await page.waitForTimeout(1000);
}

async function findVisible(page, candidates, timeoutMs = 15000) {
  const deadline = Date.now() + timeoutMs;
  while (Date.now() < deadline) {
    for (const candidate of candidates) {
      const locator = candidate.first();
      if (await locator.isVisible().catch(() => false)) {
        return locator;
      }
    }
    await page.waitForTimeout(250);
  }
  return null;
}

async function screenshot(page, field, filename, options = {}) {
  const targetPath = path.join(ARTIFACTS_DIR, filename);
  await page.screenshot({ path: targetPath, ...options });
  report[field].evidence.push(targetPath);
}

async function clickAndWait(locator, page) {
  await locator.scrollIntoViewIfNeeded().catch(() => {});
  await locator.click();
  await waitForUi(page);
}

async function validateVisibleText(page, field, label, regex, timeoutMs = 12000) {
  const locator = page.getByText(regex).first();
  try {
    await locator.waitFor({ state: "visible", timeout: timeoutMs });
    recordCheck(field, label, true);
    return locator;
  } catch (error) {
    recordCheck(field, label, false, `${label} no visible (${error.message})`);
    return null;
  }
}

async function validateEmailVisible(page, field) {
  const emailRegex = /[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}/i;
  const locator = page.getByText(emailRegex).first();
  try {
    await locator.waitFor({ state: "visible", timeout: 12000 });
    recordCheck(field, "User email is visible", true);
  } catch (error) {
    recordCheck(field, "User email is visible", false, error.message);
  }
}

async function validateLikelyUserName(page, field) {
  const nameFound = await page.evaluate(() => {
    const sections = Array.from(
      document.querySelectorAll("section, article, main, div"),
    ).filter((element) =>
      /informaci[oó]n general/i.test(element.textContent || ""),
    );

    const nameRegex = /^[A-Za-zÀ-ÿ'`.-]+(?:\s+[A-Za-zÀ-ÿ'`.-]+)+$/;
    for (const section of sections) {
      const lines = (section.innerText || "")
        .split("\n")
        .map((line) => line.trim())
        .filter(Boolean)
        .slice(0, 50);

      if (
        lines.some(
          (line) => line.length >= 4 && line.length <= 80 && nameRegex.test(line),
        )
      ) {
        return true;
      }
    }
    return false;
  });

  recordPresence(
    field,
    "User name is visible",
    nameFound,
    "No se detectó texto con formato de nombre en Información General",
  );
}

async function openLegalDocument({
  appPage,
  field,
  linkLabel,
  headingRegex,
  screenshotFile,
  metadataKey,
}) {
  const context = appPage.context();
  const legalSection = await findVisible(appPage, [
    appPage.locator("section, article, div").filter({ hasText: /secci[oó]n legal/i }),
  ]);

  if (!legalSection) {
    recordCheck(
      field,
      "Sección Legal container is visible",
      false,
      "No se encontró la Sección Legal en la vista de cuenta",
    );
    return;
  }

  const link = await findVisible(appPage, [
    legalSection.getByRole("link", { name: new RegExp(linkLabel, "i") }),
    legalSection.getByRole("button", { name: new RegExp(linkLabel, "i") }),
    legalSection.getByText(new RegExp(linkLabel, "i")),
  ]);

  if (!link) {
    recordCheck(field, `${linkLabel} link is visible`, false, "No se encontró link");
    return;
  }

  recordCheck(field, `${linkLabel} link is visible`, true);

  let legalPage = appPage;
  let popup = null;

  try {
    [popup] = await Promise.all([
      context.waitForEvent("page", { timeout: 8000 }).catch(() => null),
      clickAndWait(link, appPage),
    ]);
  } catch (error) {
    recordCheck(field, `Open ${linkLabel}`, false, error.message);
  }

  if (popup) {
    legalPage = popup;
    await legalPage.waitForLoadState("domcontentloaded", { timeout: 15000 }).catch(() => {});
    await legalPage.waitForTimeout(1000);
  }

  const heading = await findVisible(legalPage, [
    legalPage.getByRole("heading", { name: headingRegex }),
    legalPage.getByText(headingRegex),
  ]);

  recordPresence(
    field,
    `Heading ${headingRegex} is visible`,
    heading,
    `No se encontró heading ${headingRegex}`,
  );

  const legalContentFound = await legalPage.evaluate(() => {
    const paragraphLike = Array.from(
      document.querySelectorAll("p, li, article, section, div"),
    )
      .map((el) => (el.textContent || "").trim())
      .filter((text) => text.length > 100);
    return paragraphLike.length > 0;
  });

  recordPresence(
    field,
    "Legal content text is visible",
    legalContentFound,
    "No se detectó contenido legal extenso",
  );

  metadata.urls[metadataKey] = legalPage.url();
  await screenshot(legalPage, field, screenshotFile, { fullPage: true });

  if (popup) {
    await popup.close().catch(() => {});
    await appPage.bringToFront().catch(() => {});
    await waitForUi(appPage);
    return;
  }

  await appPage.goBack({ waitUntil: "domcontentloaded" }).catch(() => {});
  await waitForUi(appPage);
}

async function run() {
  await ensureArtifactsDir();

  let browser;
  let context;
  let page;
  let closeBrowser = true;

  const cdpUrl = process.env.SALEADS_CDP_URL;
  const loginUrl = process.env.SALEADS_LOGIN_URL || process.env.SALEADS_BASE_URL;

  try {
    if (cdpUrl) {
      browser = await chromium.connectOverCDP(cdpUrl);
      context = browser.contexts()[0] || (await browser.newContext());
      page = context.pages()[0] || (await context.newPage());
      closeBrowser = false;
    } else {
      browser = await chromium.launch({ headless: HEADLESS });
      context = await browser.newContext({ viewport: { width: 1440, height: 900 } });
      page = await context.newPage();

      if (!loginUrl) {
        throw new Error(
          "Falta SALEADS_LOGIN_URL o SALEADS_CDP_URL. El flujo no puede iniciar desde una URL fija.",
        );
      }

      await page.goto(loginUrl, { waitUntil: "domcontentloaded" });
      await waitForUi(page);
    }

    metadata.urls.initial = page.url();

    // Step 1: Login with Google
    const loginButton = await findVisible(page, [
      page.getByRole("button", { name: /google|sign in|iniciar sesi[oó]n|login/i }),
      page.getByRole("link", { name: /google|sign in|iniciar sesi[oó]n|login/i }),
      page.getByText(/sign in with google|iniciar sesi[oó]n con google|continuar con google|google/i),
    ]);

    recordPresence(
      "Login",
      "Login button or Google sign in is visible",
      loginButton,
      "No se encontró botón de login con Google",
    );

    if (loginButton) {
      await clickAndWait(loginButton, page);
    }

    const googleAccount = await findVisible(
      page,
      [
        page.getByText(new RegExp(GOOGLE_ACCOUNT_EMAIL, "i")),
        page.getByRole("link", { name: new RegExp(GOOGLE_ACCOUNT_EMAIL, "i") }),
        page.getByRole("button", { name: new RegExp(GOOGLE_ACCOUNT_EMAIL, "i") }),
      ],
      8000,
    );

    if (googleAccount) {
      await clickAndWait(googleAccount, page);
      recordCheck("Login", `Google account ${GOOGLE_ACCOUNT_EMAIL} selected`, true);
    }

    const sidebar = await findVisible(page, [
      page.locator("aside").filter({ hasText: /mi negocio|administrar negocios|agregar negocio/i }),
      page.locator("nav").filter({ hasText: /mi negocio|administrar negocios|agregar negocio/i }),
    ]);

    const appMain = await findVisible(page, [
      page.getByText(/mi negocio|administrar negocios|informaci[oó]n general|tus negocios/i),
      page.locator("main").filter({ hasText: /mi negocio|informaci[oó]n general|tus negocios/i }),
    ]);

    recordPresence(
      "Login",
      "Main application interface appears",
      appMain,
      "No se detectó la interfaz principal",
    );
    recordPresence(
      "Login",
      "Left sidebar navigation is visible",
      sidebar,
      "No se detectó la barra lateral izquierda",
    );
    await screenshot(page, "Login", "01-dashboard.png");

    // Step 2: Open Mi Negocio menu
    const negocioSection = await findVisible(page, [
      page.getByRole("button", { name: /^negocio$/i }),
      page.getByRole("link", { name: /^negocio$/i }),
      page.getByText(/^negocio$/i),
    ]);

    if (negocioSection) {
      await clickAndWait(negocioSection, page);
      recordCheck("Mi Negocio menu", "Negocio section found and clicked", true);
    } else {
      recordCheck("Mi Negocio menu", "Negocio section found and clicked", false);
    }

    const miNegocio = await findVisible(page, [
      page.getByRole("button", { name: /mi negocio/i }),
      page.getByRole("link", { name: /mi negocio/i }),
      page.getByText(/mi negocio/i),
    ]);

    if (miNegocio) {
      await clickAndWait(miNegocio, page);
      recordCheck("Mi Negocio menu", "Mi Negocio option clicked", true);
    } else {
      recordCheck("Mi Negocio menu", "Mi Negocio option clicked", false);
    }

    await validateVisibleText(
      page,
      "Mi Negocio menu",
      "Agregar Negocio is visible",
      /agregar negocio/i,
    );
    await validateVisibleText(
      page,
      "Mi Negocio menu",
      "Administrar Negocios is visible",
      /administrar negocios/i,
    );
    await screenshot(page, "Mi Negocio menu", "02-mi-negocio-menu-expanded.png");

    // Step 3: Validate Agregar Negocio modal
    const agregarNegocioMenuOption = await findVisible(page, [
      page.getByRole("button", { name: /agregar negocio/i }),
      page.getByRole("link", { name: /agregar negocio/i }),
      page.getByText(/agregar negocio/i),
    ]);

    if (agregarNegocioMenuOption) {
      await clickAndWait(agregarNegocioMenuOption, page);
      recordCheck("Agregar Negocio modal", "Agregar Negocio clicked", true);
    } else {
      recordCheck("Agregar Negocio modal", "Agregar Negocio clicked", false);
    }

    await validateVisibleText(
      page,
      "Agregar Negocio modal",
      "Modal title Crear Nuevo Negocio is visible",
      /crear nuevo negocio/i,
    );
    await validateVisibleText(
      page,
      "Agregar Negocio modal",
      "Nombre del Negocio input exists",
      /nombre del negocio/i,
    );
    await validateVisibleText(
      page,
      "Agregar Negocio modal",
      "Tienes 2 de 3 negocios text is visible",
      /tienes\s*2\s*de\s*3\s*negocios/i,
    );
    await validateVisibleText(
      page,
      "Agregar Negocio modal",
      "Cancelar button is present",
      /^cancelar$/i,
    );
    await validateVisibleText(
      page,
      "Agregar Negocio modal",
      "Crear Negocio button is present",
      /crear negocio/i,
    );
    await screenshot(page, "Agregar Negocio modal", "03-agregar-negocio-modal.png");

    const nombreInput = await findVisible(page, [
      page.getByLabel(/nombre del negocio/i),
      page.getByPlaceholder(/nombre del negocio/i),
      page.locator("input").filter({ hasText: /nombre del negocio/i }),
      page.locator("input[name*='negocio'], input[id*='negocio']"),
    ]);

    if (nombreInput) {
      await nombreInput.fill("Negocio Prueba Automatización");
      recordCheck(
        "Agregar Negocio modal",
        "Optional input typing completed",
        true,
      );
    }

    const cancelarButton = await findVisible(page, [
      page.getByRole("button", { name: /^cancelar$/i }),
      page.getByText(/^cancelar$/i),
    ]);
    if (cancelarButton) {
      await clickAndWait(cancelarButton, page);
      recordCheck("Agregar Negocio modal", "Modal closed with Cancelar", true);
    }

    // Step 4: Open Administrar Negocios
    const miNegocioForAdmin = await findVisible(page, [
      page.getByRole("button", { name: /mi negocio/i }),
      page.getByRole("link", { name: /mi negocio/i }),
      page.getByText(/mi negocio/i),
    ]);
    if (miNegocioForAdmin) {
      await clickAndWait(miNegocioForAdmin, page);
    }

    const administrarNegocios = await findVisible(page, [
      page.getByRole("button", { name: /administrar negocios/i }),
      page.getByRole("link", { name: /administrar negocios/i }),
      page.getByText(/administrar negocios/i),
    ]);

    if (administrarNegocios) {
      await clickAndWait(administrarNegocios, page);
      recordCheck("Administrar Negocios view", "Administrar Negocios clicked", true);
    } else {
      recordCheck("Administrar Negocios view", "Administrar Negocios clicked", false);
    }

    await validateVisibleText(
      page,
      "Administrar Negocios view",
      "Información General section exists",
      /informaci[oó]n general/i,
    );
    await validateVisibleText(
      page,
      "Administrar Negocios view",
      "Detalles de la Cuenta section exists",
      /detalles de la cuenta/i,
    );
    await validateVisibleText(
      page,
      "Administrar Negocios view",
      "Tus Negocios section exists",
      /tus negocios/i,
    );
    await validateVisibleText(
      page,
      "Administrar Negocios view",
      "Sección Legal section exists",
      /secci[oó]n legal/i,
    );
    await screenshot(page, "Administrar Negocios view", "04-administrar-negocios.png", {
      fullPage: true,
    });

    // Step 5: Validate Información General
    await validateLikelyUserName(page, "Información General");
    await validateEmailVisible(page, "Información General");
    await validateVisibleText(
      page,
      "Información General",
      "BUSINESS PLAN is visible",
      /business plan/i,
    );
    await validateVisibleText(
      page,
      "Información General",
      "Cambiar Plan button is visible",
      /cambiar plan/i,
    );

    // Step 6: Validate Detalles de la Cuenta
    await validateVisibleText(
      page,
      "Detalles de la Cuenta",
      "Cuenta creada is visible",
      /cuenta creada/i,
    );
    await validateVisibleText(
      page,
      "Detalles de la Cuenta",
      "Estado activo is visible",
      /estado activo/i,
    );
    await validateVisibleText(
      page,
      "Detalles de la Cuenta",
      "Idioma seleccionado is visible",
      /idioma seleccionado/i,
    );

    // Step 7: Validate Tus Negocios
    await validateVisibleText(
      page,
      "Tus Negocios",
      "Business list is visible",
      /tus negocios/i,
    );
    await validateVisibleText(
      page,
      "Tus Negocios",
      "Agregar Negocio button exists",
      /agregar negocio/i,
    );
    await validateVisibleText(
      page,
      "Tus Negocios",
      "Tienes 2 de 3 negocios text is visible",
      /tienes\s*2\s*de\s*3\s*negocios/i,
    );

    // Step 8: Validate Términos y Condiciones
    await openLegalDocument({
      appPage: page,
      field: "Términos y Condiciones",
      linkLabel: "Términos y Condiciones",
      headingRegex: /t[eé]rminos y condiciones/i,
      screenshotFile: "05-terminos-y-condiciones.png",
      metadataKey: "terminosYCondicionesUrl",
    });

    // Step 9: Validate Política de Privacidad
    await openLegalDocument({
      appPage: page,
      field: "Política de Privacidad",
      linkLabel: "Política de Privacidad",
      headingRegex: /pol[ií]tica de privacidad/i,
      screenshotFile: "06-politica-de-privacidad.png",
      metadataKey: "politicaDePrivacidadUrl",
    });

    metadata.urls.finalAppPage = page.url();
  } finally {
    if (browser && closeBrowser) {
      await browser.close();
    }
  }

  metadata.finishedAt = new Date().toISOString();
  finalizeFieldStatuses();
  const overallPass = REPORT_FIELDS.every((field) => report[field].status === "PASS");

  const finalPayload = {
    name: "saleads_mi_negocio_full_test",
    overallStatus: overallPass ? "PASS" : "FAIL",
    report,
    metadata,
  };

  const reportPath = path.join(ARTIFACTS_DIR, "final-report.json");
  await fs.writeFile(reportPath, `${JSON.stringify(finalPayload, null, 2)}\n`, "utf8");

  console.log(JSON.stringify(finalPayload, null, 2));
  process.exitCode = overallPass ? 0 : 1;
}

run().catch(async (error) => {
  finalizeFieldStatuses();
  const failedPayload = {
    name: "saleads_mi_negocio_full_test",
    overallStatus: "FAIL",
    fatalError: error.message,
    report,
    metadata: {
      ...metadata,
      finishedAt: new Date().toISOString(),
    },
  };

  await ensureArtifactsDir();
  await fs.writeFile(
    path.join(ARTIFACTS_DIR, "final-report.json"),
    `${JSON.stringify(failedPayload, null, 2)}\n`,
    "utf8",
  );
  console.error(JSON.stringify(failedPayload, null, 2));
  process.exitCode = 1;
});
